modules/openapi-generator/src/main/resources/typescript-fetch/api.mustache

```diff
@@ -1,364 +1,442 @@
-/// <reference path="./custom.d.ts" />
 // tslint:disable
 {{>licenseInfo}}
 
-import * as url from "url";
-import * as portableFetch from "portable-fetch";
-import { Configuration } from "./configuration";
+export const BASE_PATH = "{{{basePath}}}".replace(/\/+$/, "");
 
-const BASE_PATH = "{{{basePath}}}".replace(/\/+$/, "");
-
-/**
- *
- * @export
- */
-export const COLLECTION_FORMATS = {
-    csv: ",",
-    ssv: " ",
-    tsv: "\t",
-    pipes: "|",
-};
+{{#models}}
+{{#model}}{{#isEnum}}{{>modelEnum}}{{/isEnum}}{{^isEnum}}{{>modelGeneric}}{{/isEnum}}{{/model}}
+{{/models}}
 
-/**
- *
- * @export
- * @interface FetchAPI
- */
-export interface FetchAPI {
-    (url: string, init?: any): Promise<Response>;
+{{#apiInfo}}
+{{#apis}}
+{{#operations}}
+{{#operation}}
+{{#allParams.0}}
+export interface {{operationIdCamelCase}}Request {
+    {{#allParams}}
+    {{paramName}}{{^required}}?{{/required}}: {{{dataType}}};
+    {{/allParams}}
 }
 
-/**
- *  
- * @export
- * @interface FetchArgs
- */
-export interface FetchArgs {
-    url: string;
-    options: any;
-}
+{{/allParams.0}}
+{{/operation}}
+{{/operations}}
+{{/apis}}
+{{/apiInfo}}
 
 /**
- * 
- * @export
- * @class BaseAPI
+ * This is the base class for all generated API classes.
  */
 export class BaseAPI {
-    protected configuration: Configuration;
 
-    constructor(configuration?: Configuration, protected basePath: string = BASE_PATH, protected fetch: FetchAPI = portableFetch) {
-        if (configuration) {
-            this.configuration = configuration;
-            this.basePath = configuration.basePath || this.basePath;
+    private middleware: Middleware[];
+
+    constructor(protected configuration = new Configuration()) {
+        this.middleware = configuration.middleware;
+    }
+
+    withMiddleware<T extends BaseAPI>(this: T, ...middlewares: Middleware[]) {
```

Comment:
@TiFu our generator supports middleware. This is a very useful feature and something that most HTTP clients contain. My colleagues have been able to make good use of this feature across different projects to implement logging, automatic request retries, timeouts, circuit breakers, automatic token refresh logic, automatic logout logic, complex authentication/authorization challenges that are custom or out-of-scope for the built in security definitions, etc.